File: packages_ts/backend/clarity-core/src/model/model.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { Fact } from '@relica/types';

import { ArchivistService } from '../archivist/archivist.service.js';

const physObjUID = 730044;
const aspectUID = 790229;
const roleUID = 160170;
const relationUID = 2850;
const occurrenceUID = 193671;

const PHYSICAL_OBJECT = 'physical object';
const ASPECT = 'aspect';
const ROLE = 'role';
const RELATION = 'relation';
const OCCURRENCE = 'occurrence';

@Injectable()
export class ModelService {
  private readonly logger = new Logger(ModelService.name);

  constructor(private readonly archivistService: ArchivistService) {}

  async getPhysicalObjectModel(uid) {
    return { aspects: [], roles: [], components: [], connections: [] };
  }

  async getAspectModel(uid) {
    return { possessors: [] };
  }

  async getRoleModel(uid) {
    return { rolePlayers: [] };
  }

  async getRelationModel(uid) {
    return { rolePlayer1: null, rolePlayer2: null };
  }

  async getOccurrenceModel(uid) {
    return { aspects: [], involved: [] };
  }

  async retrieveKindModel(uid, token) {
    console.log('RETREIVE MODEL uid', uid);
    const category = await this.archivistService.getCategory(uid, token);
    console.log('category', category);
    const facts = await this.archivistService.retrieveAllFacts(uid, token);
    // console.log('facts', facts);
    const definitiveFacts = await this.archivistService.getDefinitiveFacts(
      uid,
      token,
    );
    // console.log('definitiveFacts', definitiveFacts);
    const specialization =
      await this.archivistService.getRelatedOnUIDSubtypeCone(uid, 1146, token);
    // console.log('specialization', specialization);
    const classification =
      await this.archivistService.getRelatedOnUIDSubtypeCone(uid, 1225, token);
    // console.log("classification", classification);
    const synonyms = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      1981,
      token,
    );
    // console.log("synonyms", synonyms);
    const inverses = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      1986,
      token,
    );
    // console.log("inverses", inverses);
    const reqRole1 = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      4731,
      token,
    );
    // console.log("reqRole1", reqRole1);
    const reqRole2 = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      4733,
      token,
    );
    // console.log("reqRole2", reqRole2);
    const possRoles = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      4714,
      token,
    );
    // console.log("possRoles", possRoles);
    let model;
    switch (category) {
      case PHYSICAL_OBJECT:
        model = await this.getPhysicalObjectModel(uid);
        break;
      case ASPECT:
        model = await this.getAspectModel(uid);
        break;
      case ROLE:
        model = await this.getRoleModel(uid);
        break;
      case RELATION:
        model = await this.getRelationModel(uid);
        break;
      case OCCURRENCE:
        model = await this.getOccurrenceModel(uid);
        break;
      default:
        model = {};
        break;
    }

    // console.log(definitiveFacts);
    return Object.assign(model, {
      uid: uid,
      collection: {
        uid: definitiveFacts[0].collection_uid,
        name: definitiveFacts[0].collection_name,
      },
      name: definitiveFacts[0].lh_object_name, //.map((x: Fact)  x.lh_object_name).join(", "),
      type: 'kind',
      category: category,
      definition: definitiveFacts.map((x) => ({
        fact_uid: x.fact_uid,
        partial_definition: x.partial_definition,
        full_definition: x.full_definition,
      })),
      facts: facts,
      //
      1146: specialization.map((x) => x.rh_object_uid),
      1225: classification.map((x) => x.rh_object_uid),
      1981: synonyms.map((x) => x.lh_object_name),
      1986: inverses.map((x) => x.lh_object_name),
      4731: reqRole1.map((x) => x.rh_object_uid),
      4733: reqRole2.map((x) => x.rh_object_uid),
      4714: possRoles.map((x) => x.rh_object_uid),
    });
  }

  async retrieveIndividualModel(uid: number, token: string) {
    const category = await this.archivistService.getCategory(uid, token);
    const facts = await this.archivistService.retrieveAllFacts(uid, token);
    const definitiveFacts = await this.archivistService.getDefinitiveFacts(
      uid,
      token,
    );
    const classification =
      await this.archivistService.getRelatedOnUIDSubtypeCone(uid, 1225, token);
    // // const synonyms = await getRelatedOnUIDSubtypeCone(uid, 1981);
    // // const inverses = await getRelatedOnUIDSubtypeCone(uid, 1986);
    // // const reqRole1 = await getRelatedOnUIDSubtypeCone(uid, 4731);
    // // const reqRole2 = await getRelatedOnUIDSubtypeCone(uid, 4733);
    // // const possRoles = await getRelatedOnUIDSubtypeCone(uid, 4714);
    if (classification.length === 0) {
      return null;
    }
    const baseObj = {
      uid: uid,
      name: classification[0].lh_object_name,
      collection: {
        uid: definitiveFacts[0].collection_uid,
        name: definitiveFacts[0].collection_name,
      },
      1225: classification.map((x) => x.rh_object_uid),
      type: 'individual',
      category: category,
      definition: definitiveFacts.map((x) => ({
        fact_uid: x.fact_uid,
        partial_definition: x.partial_definition,
        full_definition: x.full_definition,
      })),
      facts,
    };
    const value = await this.archivistService.getRelatedOnUIDSubtypeCone(
      uid,
      5025,
      token,
    ); // 'has on scale a value equal to'
    if (value.length > 0) {
      const valFact = value[0];
      const val = parseInt(valFact.rh_object_name);
      const uom = { uid: valFact.uom_uid, name: valFact.uom_name };
      return { ...baseObj, value: { quant: val, uom } };
    }
    return baseObj;
  }

  async retrieveQualificationModel(uid: number, token: string) {
    const category = await this.archivistService.getCategory(uid, token);
    const facts = await this.archivistService.retrieveAllFacts(uid, token);
    const definitiveFacts = await this.archivistService.getDefinitiveFacts(
      uid,
      token,
    );
    const baseObj = {
      name: definitiveFacts[0].lh_object_name,
      uid: uid,
      type: 'qualification',
      category: category,
      facts,
    };
    return baseObj;
  }

  async retrieveModel(uid: number, token: string) {
    const type = await this.archivistService.getEntityType(uid, token);

    console.log('type', type);

    if (type === 'kind') {
      return this.retrieveKindModel(uid, token);
    } else if (type === 'individual') {
      return this.retrieveIndividualModel(uid, token);
    } else if (type === 'qualification') {
      return this.retrieveQualificationModel(uid, token);
    } else if (uid === 730000) {
      return {
        uid: uid,
        name: 'anything',
        type: 'kind',
        category: 'anything',
        definition: 'is an anything',
        facts: [],
        1146: [],
        1225: [],
        1981: [],
        1986: [],
        4731: [],
        4733: [],
        4714: [],
      };
    } else {
      console.log(
        '//// ERROR: modelController.retrieveModel: unknown type uid',
        uid,
      );
      return null;
    }
  }

  async throttlePromises(funcs, limit) {
    let results = [];
    let executing = [];
    for (const func of funcs) {
      const p = Promise.resolve().then(func);
      results.push(p);
      if (limit <= funcs.length) {
        const e = p.then(() => executing.splice(executing.indexOf(e), 1));
        executing.push(e);
        if (executing.length >= limit) {
          await Promise.race(executing);
        }
      }
    }
    return Promise.all(results);
  }

  async retrieveModels(uids: number[], token) {
    const funcs = uids.map((uid) => () => this.retrieveModel(uid, token));
    return await this.throttlePromises(funcs, 5);
  }

  async updateDefinition(fact_uid, partial_definition, full_definition) {
    const response = await this.archivistService.submitDefinition(
      fact_uid,
      partial_definition,
      full_definition,
    );
    return response;
  }

  async updateCollection(fact_uid, collection_uid, collection_name) {
    const response = await this.archivistService.submitCollection(
      fact_uid,
      collection_uid,
      collection_name,
    );
    return response;
  }

  async updateName(uid, name) {
    // fact_uid, (the definitive fact uid)
    // new_name
    const response = await this.archivistService.submitName(uid, name);
    return response;
  }
}

File: packages_clj/clarity/src/rlc/clarity/services/model_service.clj
(ns rlc.clarity.services.model-service
  (:require [clojure.tools.logging :as log]
            [clojure.core.async :refer [go <!]]
            [io.relica.common.io.archivist-client :as archivist]
            [rlc.clarity.io.client-instances :refer [archivist-client]]))

;; Constants from the TypeScript version
(def phys-obj-uid 730044)
(def aspect-uid 790229)
(def role-uid 160170)
(def relation-uid 2850)
(def occurrence-uid 193671)

(def physical-object "physical object")
(def aspect "aspect")
(def role "role")
(def relation "relation")
(def occurrence "occurrence")

;; Helper functions
;; (defn get-entity-type
;;   "Get the entity type for a given UID"
;;   [uid]
;;   (client/make-request "/retrieveEntity/type"
;;                       {:query-params {:uid uid}
;;                        :throw-exceptions false}))


(defn get-category [uid]
  (go
    (try
      (let [response (<! (archivist/get-entity-category
                          archivist-client
                          uid))]
        (log/info (str "Response from get-entity-category: " response))
        (if (:success response)
          (:category response) ;; (success-response (:results response))
          (or (:error response) "Unknown error")))
      (catch Exception e
        (tap> (str "Error in text search handler: " e))
        "Failed to execute text search"))))

(defn retrieve-all-facts
  "Retrieve all facts for a given UID"
  [uid]
  (go
    (try
      (let [response (<! (archivist/get-all-related
                          archivist-client
                          uid))]
        (log/info (str "Response from get-all-facts: " response))
        (if (:success response)
          (:facts response)
          (or (:error response) "Unknown error")))
      (catch Exception e
        (tap> (str "Error in text search handler: " e))
        "Failed to execute text search"))))


(defn get-definitive-facts
  "Get definitive facts for a given UID"
  [uid]
  (go
    (try
      (let [response (<! (archivist/get-definitive-facts
                          archivist-client
                          uid))]
        (log/info (str "Response from get-definitive-facts: " response))
        (if (:success response)
          (:facts response)
          (or (:error response) "Unknown error")))
      (catch Exception e
        (tap> (str "Error in text search handler: " e))
        "Failed to execute text search"))))

  ;; (client/make-request "/fact/definitiveFacts"
  ;;                     {:query-params {:uid uid}
  ;;                      :throw-exceptions false}))

;; (defn get-related-on-uid-subtype-cone
;;   "Get related entities on UID subtype cone"
;;   [lh-object-uid rel-type-uid]
;;   (client/make-request "/fact/relatedOnUIDSubtypeCone"
;;                       {:query-params {:lh_object_uid lh-object-uid
;;                                      :rel_type_uid rel-type-uid}
;;                        :throw-exceptions false}))

;; Model retrieval functions

(defn get-physical-object-model
  "Get physical object model for a given UID"
  [uid]
  {:aspects [] :roles [] :components [] :connections []})

(defn get-aspect-model
  "Get aspect model for a given UID"
  [uid]
  {:possessors []})

(defn get-role-model
  "Get role model for a given UID"
  [uid]
  {:role-players []})

(defn get-relation-model
  "Get relation model for a given UID"
  [uid]
  {:role-player1 nil :role-player2 nil})

(defn get-occurrence-model
  "Get occurrence model for a given UID"
  [uid]
  {:aspects [] :involved []})

(defn retrieve-kind-model
  "Retrieve kind model for a given UID"
  [uid]
  (log/info "Retrieving kind model for UID:" uid)
  (go
    (let [category (<! (get-category uid))
          _ (log/info  (str "Category:" category))
          facts (<! (retrieve-all-facts uid))
          definitive-facts (<! (get-definitive-facts uid))
          ;; specialization (get-related-on-uid-subtype-cone uid 1146)
          ;; classification (get-related-on-uid-subtype-cone uid 1225)
          ;; synonyms (get-related-on-uid-subtype-cone uid 1981)
          ;; inverses (get-related-on-uid-subtype-cone uid 1986)
          ;; req-role1 (get-related-on-uid-subtype-cone uid 4731)
          ;; req-role2 (get-related-on-uid-subtype-cone uid 4733)
          ;; poss-roles (get-related-on-uid-subtype-cone uid 4714)
          base-model (case category
                       "physical object" (get-physical-object-model uid)
                       "aspect" (get-aspect-model uid)
                       "role" (get-role-model uid)
                       "relation" (get-relation-model uid)
                       "occurrence" (get-occurrence-model uid)
                       {})
          ]
      (log/info "category:" category)
      (log/info "Base model:" base-model)
      (log/info "Facts:" facts)
      ;; (when (seq definitive-facts)
      (merge base-model
             {:uid uid
    ;;           :collection {:uid (get-in definitive-facts [0 :collection_uid])
    ;;                       :name (get-in definitive-facts [0 :collection_name])}
    ;;           :name (get-in definitive-facts [0 :lh_object_name])
              :type "kind"
              :category category
              :definition (mapv (fn [x]
                                 {:fact_uid (:fact_uid x)
                                  :partial_definition (:partial_definition x)
                                  :full_definition (:full_definition x)})
                               definitive-facts)
              :facts facts
    ;;           ;; Relation UIDs as keys
    ;;           1146 (mapv #(get % :rh_object_uid) specialization)
    ;;           1225 (mapv #(get % :rh_object_uid) classification)
    ;;           1981 (mapv #(get % :lh_object_name) synonyms)
    ;;           1986 (mapv #(get % :lh_object_name) inverses)
    ;;           4731 (mapv #(get % :rh_object_uid) req-role1)
    ;;           4733 (mapv #(get % :rh_object_uid) req-role2)
    ;;           4714 (mapv #(get % :rh_object_uid) poss-roles)
              }
             )
    ;;           )
    )))

(defn retrieve-individual-model
  "Retrieve individual model for a given UID"
  [uid]
  (log/info "Retrieving individual model for UID:" uid)
  (go
    (let [category (<! (get-category uid))
          facts (<! (retrieve-all-facts uid))
          definitive-facts (<! (get-definitive-facts uid))
    ;;     classification (get-related-on-uid-subtype-cone uid 1225)
        ]
    ;; (when (and (seq classification) (seq definitive-facts))
      (let [base-obj {:uid uid
    ;;                  :name (get-in classification [0 :lh_object_name])
                      :collection {:uid 0 :name ""}
    ;;                  :collection {:uid (get-in definitive-facts [0 :collection_uid])
    ;;                              :name (get-in definitive-facts [0 :collection_name])}
    ;;                  1225 (mapv #(get % :rh_object_uid) classification)
                     :type "individual"
                     :category category
                     ;; :definition ["is an individual"]
                     :definition (mapv (fn [x]
                                        {:fact_uid (:fact_uid x)
                                         :partial_definition (:partial_definition x)
                                         :full_definition (:full_definition x)})
                                      definitive-facts)
                     :facts facts
                     }
    ;;         value (get-related-on-uid-subtype-cone uid 5025)
                    ] ; 'has on scale a value equal to'
    ;;     (if (seq value)
    ;;       (let [val-fact (first value)
    ;;             val (Integer/parseInt (:rh_object_name val-fact))
    ;;             uom {:uid (:uom_uid val-fact) :name (:uom_name val-fact)}]
    ;;         (assoc base-obj :value {:quant val :uom uom}))
        base-obj
        )
    ;;       )
      )))

;; (defn retrieve-qualification-model
;;   "Retrieve qualification model for a given UID"
;;   [uid]
;;   (log/info "Retrieving qualification model for UID:" uid)
;;   (let [category (get-category uid)
;;         facts (retrieve-all-facts uid)
;;         definitive-facts (get-definitive-facts uid)]
;;     (when (seq definitive-facts)
;;       {:name (get-in definitive-facts [0 :lh_object_name])
;;        :uid uid
;;        :type "qualification"
;;        :category category
;;        :facts facts})))

;; (defn retrieve-model
;;   "Retrieve model for a given UID"
;;   [uid]
;;   (log/info "Retrieving model for UID:" uid)
;;   (let [type (get-entity-type uid)]
;;     (log/debug "Entity type:" type)
;;     (cond
;;       (= type "kind") (retrieve-kind-model uid)
;;       (= type "individual") (retrieve-individual-model uid)
;;       (= type "qualification") (retrieve-qualification-model uid)
;;       (= uid 730000) {:uid uid
;;                      :name "anything"
;;                      :type "kind"
;;                      :category "anything"
;;                      :definition "is an anything"
;;                      :facts []
;;                      1146 []
;;                      1225 []
;;                      1981 []
;;                      1986 []
;;                      4731 []
;;                      4733 []
;;                      4714 []}
;;       :else (do
;;               (log/error "Unknown entity type for UID:" uid)
;;               nil))))

;; (defn throttle-promises
;;   "Execute a sequence of functions with a limit on concurrent executions"
;;   [funcs limit]
;;   (let [results (atom [])
;;         executing (atom [])
;;         promise-all (promise)]
;;     (doseq [func funcs]
;;       (let [p (future (func))]
;;         (swap! results conj p)
;;         (when (<= limit (count funcs))
;;           (let [e (future
;;                     (deref p)
;;                     (swap! executing #(remove #{e} %)))]
;;             (swap! executing conj e)
;;             (when (>= (count @executing) limit)
;;               (deref (first @executing)))))))
;;     (future
;;       (doseq [r @results]
;;         (deref r))
;;       (deliver promise-all @results))
;;     @promise-all))

;; (defn retrieve-models
;;   "Retrieve models for a sequence of UIDs"
;;   [uids]
;;   (log/info "Retrieving models for UIDs:" uids)
;;   (let [funcs (map (fn [uid] #(retrieve-model uid)) uids)]
;;     (throttle-promises funcs 5)))

;; (defn update-definition
;;   "Update definition for a fact"
;;   [fact-uid partial-definition full-definition]
;;   (log/info "Updating definition for fact:" fact-uid)
;;   (client/make-put-request "/submission/definition"
;;                           {:fact_uid fact-uid
;;                            :partial_definition partial-definition
;;                            :full_definition full-definition}
;;                           {:throw-exceptions false}))

;; (defn update-collection
;;   "Update collection for a fact"
;;   [fact-uid collection-uid collection-name]
;;   (log/info "Updating collection for fact:" fact-uid)
;;   (client/make-put-request "/submission/collection"
;;                           {:fact_uid fact-uid
;;                            :collection_uid collection-uid
;;                            :collection_name collection-name}
;;                           {:throw-exceptions false}))

;; (defn update-name
;;   "Update name for a fact"
;;   [uid name]
;;   (log/info "Updating name for fact:" uid)
;;   (client/make-put-request "/submission/name"
;;                           {:fact_uid uid
;;                            :name name}
;;                           {:throw-exceptions false}))

;; Service creation
(defn create-model-service
  "Create a new model service"
  []
   {
  ;; :retrieve-model retrieve-model
  ;;  :retrieve-models retrieve-models
   :retrieve-kind-model retrieve-kind-model
   :retrieve-individual-model retrieve-individual-model
   ;; :retrieve-qualification-model retrieve-qualification-model
   ;; :update-definition update-definition
   ;; :update-collection update-collection
   ;;:update-name update-name
   })
